ledger/sandbox-on-x/src/main/scala/com/daml/ledger/sandbox/bridge/ConflictCheckingLedgerBridge.scala

```diff
@@ -377,56 +378,49 @@ private[sandbox] class ConflictCheckingLedgerBridge(
               .map { lookupResult =>
                 (inputState, lookupResult) match {
                   case (LfTransaction.NegativeKeyLookup, Some(actual)) =>
-                    Left(InconsistentContractKey(None, Some(actual))(errorFactories))
+                    Left(
+                      InconsistentContractKey(None, Some(actual))(completionInfo, errorFactories)
+                    )
                   case (LfTransaction.KeyCreate, Some(_)) =>
-                    Left(DuplicateKey(key)(errorFactories))
+                    Left(DuplicateKey(key)(completionInfo, errorFactories))
                   case (LfTransaction.KeyActive(expected), actual) if !actual.contains(expected) =>
-                    Left(InconsistentContractKey(Some(expected), actual)(errorFactories))
+                    Left(
+                      InconsistentContractKey(Some(expected), actual)(
+                        completionInfo,
+                        errorFactories,
+                      )
+                    )
                   case _ => Right(())
                 }
               }
-              .map(
-                _.left.map(toCommandRejectedUpdate(_, transaction.submitterInfo.toCompletionInfo()))
-              )
           case left => Future.successful(left)
         }
     }
+  }
 
-  private def toCommandRejectedUpdate(rejection: Rejection, completionInfo: CompletionInfo) =
-    Update.CommandRejected(
-      recordTime = timeProvider.getCurrentTimestamp,
-      completionInfo = completionInfo,
-      reasonTemplate = FinalReason(rejection.toStatus),
-    )
-
-  private def invalidInputFromParticipantRejection(
-      completionInfo: CompletionInfo
-  )(implicit
+  private def invalidInputFromParticipantRejection(completionInfo: CompletionInfo)(implicit
       contextualizedErrorLogger: ContextualizedErrorLogger
-  ): LfTransaction.KeyInputError => Update.CommandRejected = {
```

Comment:
`recordTime` is captured in `sequence` together with offsets assignment. Then, let's generate plain rejections on earlier stages and generate the CommandRejected as the last step in the `sequence` stage.